File: android/app/src/main/java/com/rcforb/android/protocol/CommandParser.kt
package com.rcforb.android.protocol

object CommandParser {
    fun splitCommand(command: String): List<String> = command.split("::")

    fun getKey(data: String): String {
        val idx = data.indexOf("::")
        return if (idx >= 0) data.substring(0, idx) else data
    }

    fun getValue(data: String): String {
        val parts = splitCommand(data)
        return if (parts.size > 1) parts[1] else ""
    }

    fun getValueAt(index: Int, data: String): String {
        val parts = splitCommand(data)
        return if (index < parts.size) parts[index] else ""
    }

    // Command builders
    fun loginCmd(user: String, passwordMD5: String): String = "login $user $passwordMD5"
    fun setProtocolRCS(): String = "set protocol rcs"
    fun requestRadioState(): String = "radio::request-state"

    fun setFrequencyA(hz: String): String = "radio::frequency::$hz"
    fun setFrequencyB(hz: String): String = "radio::frequencyb::$hz"
    fun setButton(name: String, value: String): String = "radio::button::$name::$value"
    fun setDropdown(name: String, value: String): String = "radio::dropdown::$name::$value"
    fun setSlider(name: String, value: String): String = "radio::slider::$name::$value"
    fun setMessage(name: String, value: String): String = "radio::message::$name::$value"
    fun chatMessage(text: String): String = "post::chat::$text"
    fun heartbeatCmd(): String = "post::heartbeat::${System.currentTimeMillis()}"
    fun checkCanTune(): String = "post::check::cantune"

    fun rotatorBearing(value: String): String = "rotator::bearing::$value"
    fun rotatorElevation(value: String): String = "rotator::elevation::$value"
    fun rotatorStart(): String = "rotator::start"
    fun rotatorStop(): String = "rotator::stop"
    fun ampButton(name: String, value: String): String = "amp::button::$name::$value"
    fun switchButton(name: String, value: String): String = "switch::button::$name::$value"
}

File: macOS/RCForb/Sources/Protocol/CommandParser.swift
import Foundation

enum CommandParser {
    static func splitCommand(_ command: String) -> [String] {
        command.components(separatedBy: "::")
    }

    static func getKey(_ data: String) -> String {
        if let idx = data.range(of: "::") {
            return String(data[data.startIndex..<idx.lowerBound])
        }
        return data
    }

    static func getValue(_ data: String) -> String {
        let parts = splitCommand(data)
        return parts.count > 1 ? parts[1] : ""
    }

    static func getValueAt(_ index: Int, _ data: String) -> String {
        let parts = splitCommand(data)
        return index < parts.count ? parts[index] : ""
    }

    // Command builders
    static func loginCmd(_ user: String, _ passwordMD5: String) -> String {
        "login \(user) \(passwordMD5)"
    }

    static func setProtocolRCS() -> String { "set protocol rcs" }
    static func requestRadioState() -> String { "radio::request-state" }

    static func setFrequencyA(_ hz: String) -> String { "radio::frequency::\(hz)" }
    static func setFrequencyB(_ hz: String) -> String { "radio::frequencyb::\(hz)" }
    static func setButton(_ name: String, _ value: String) -> String { "radio::button::\(name)::\(value)" }
    static func setDropdown(_ name: String, _ value: String) -> String { "radio::dropdown::\(name)::\(value)" }
    static func setSlider(_ name: String, _ value: String) -> String { "radio::slider::\(name)::\(value)" }
    static func setMessage(_ name: String, _ value: String) -> String { "radio::message::\(name)::\(value)" }
    static func chatMessage(_ text: String) -> String { "post::chat::\(text)" }
    static func heartbeatCmd() -> String { "post::heartbeat::\(Int(Date().timeIntervalSince1970 * 1000))" }
    static func checkCanTune() -> String { "post::check::cantune" }

    static func rotatorBearing(_ value: String) -> String { "rotator::bearing::\(value)" }
    static func rotatorElevation(_ value: String) -> String { "rotator::elevation::\(value)" }
    static func rotatorStart() -> String { "rotator::start" }
    static func rotatorStop() -> String { "rotator::stop" }
    static func ampButton(_ name: String, _ value: String) -> String { "amp::button::\(name)::\(value)" }
    static func switchButton(_ name: String, _ value: String) -> String { "switch::button::\(name)::\(value)" }
}
